lib/Caxy/HtmlDiff/ListDiff.php

```diff
@@ -0,0 +1,366 @@
+        }
+        // Return the result.
+        return implode(' ', $returnText);
+    }
+    
+    protected function checkWordForDiffTag($word)
+    {
+        foreach ($this->isolatedDiffTags as $diffTag) {
+            if (strpos($word, $diffTag) > -1) {
+                $position = strpos($word, $diffTag);
+                $length = strlen($diffTag);
+                $result = array(
+                    substr($word, 0, $position),
+                    $diffTag,
+                    substr($word, ($position + $length))
+                );
+                
+                return $result;
+            }
+        }
+        
+        return $word;
+    }
+    
+    /**
+     * Used to remove new lines.
+     */
+    protected function stripNewLine($text)
+    {
+        return trim(preg_replace('/\s\s+/', ' ', $text));
+    }
+    
+    /**
+     * Grab the list content using the listsIndex array.
+     */
+    protected function getListContent($indexKey = 'new', array $matches)
+    {
+        $bucket = array();
+        $start = $this->listsIndex[$indexKey][$matches[$indexKey]];
+        $stop = $this->listsIndex[$indexKey][$matches[$indexKey] + 1];
+        for ($x = $start; $x < $stop; $x++) {
+            if (in_array($this->list[$indexKey][$x], $this->isolatedDiffTags)) {
+                $bucket[] = $this->listIsolatedDiffTags[$indexKey][$x]; 
+            }
+        }
+        return $bucket;
+    }
+    
+    /**
+     * indexLists
+     * 
+     * Index the list, starting positions, so that we can refer back to it later.
+     * This is used to see where one list node starts and another ends.
+     */
+    protected function indexLists()
+    {
+        $this->listsIndex = array();
+        $lookingFor = "<li>";
+        
+        foreach ($this->list as $type => $list) {
+            $this->listsIndex[$type] = array();
+            
+            foreach ($list as $key => $listItem) {
+                if ($listItem == $lookingFor) {
+                    $this->listsIndex[$type][] = $key;
+                }
+            }
+        }
+    }
+    
+    /**
+     * Adds the opening or closing list html element, based on listType.
+     */
+    protected function addListTypeWrapper($opening = true)
+    {
+        return "<" . (!$opening ? "/" : '') . $this->listType . ">";
+    }
+    
+    /**
+     * Replace nested list with placeholders.
+     */
+    public function replaceListIsolatedDiffTags()
+    {
+        $this->listIsolatedDiffTags['old'] = $this->createIsolatedDiffTagPlaceholders($this->list['old']);
+        $this->listIsolatedDiffTags['new'] = $this->createIsolatedDiffTagPlaceholders($this->list['new']);
+    }
+    
+    /**
+     * Grab the contents of a list node.
+     */
+    protected function getListsContent(array $contentArray, $stripTags = true)
+    {
+        preg_match_all('/<li>(.*?)<\/li>/s', implode('', $contentArray), $matches);
```

Comment:
This regex doesn't account for list items that would have attributes like class or id. I also advise against doing a regex match like this because we can't ignore the chances of getting HTML that has nested `<li>` tags without an ol, dl, or ul. For example:

```html
<ul>
    <li>List item with <li>nested list item</li> and some content afterwards</li>
</ul>
```
In this case, it would match: `<li>List item with <li>nested list item</li>` and would miss the closing list item.

I think we'll want to either:
a. loop through the contentArray and keep track of when we're inside a `<li>` and count how many nested `<li>`'s are encountered in order to find the matching closing `</li>` (this is how `createIsolatedDiffTagPlaceholders` function works).
OR
b. use [DOMDocument](http://php.net/manual/en/class.domdocument.php) classes to process the HTML. I think this might be slower than option a (?) but that's just a guess. Using DOMDocument classes could simplify a lot of the HTML processing of lists, but might be overkill for what we're trying to do here.